platform-hub-api/app/models/kubernetes_token.rb

```diff
@@ -1,18 +1,142 @@
-class KubernetesToken < KubernetesTokenBase
-  attr_accessor :identity_id, :expire_privileged_at
+class KubernetesToken < ApplicationRecord
+  TOKEN_LENGTH = 36
+  UID_LENGTH = 36
+  PRIVILEGED_GROUP_MAX_EXPIRATION_SECONDS = 21600 # max privilege period is 6h
 
-  validates_presence_of :identity_id
+  NAME_REGEX = /\A[a-zA-Z][\@\.\w-]*\z/
 
-  belongs_to :identity
+  include Audited
+  audited descriptor_field: :name, associated_field: :tokenable
 
-  def self.from_data(attributes)
-    attributes = attributes.with_indifferent_access
-    params = attributes
-      .extract!(:identity_id, :cluster, :uid, :groups, :expire_privileged_at)
-      .merge!(token: ENCRYPTOR.decrypt(attributes[:token]))
-      .with_indifferent_access
+  attr_readonly :token, :name, :uid, :kind, :cluster_id
```

Comment:
I guess we should be able to update the `name`? (sorry, I know I probably suggested making it readonly in the first place!)

Reply:
There was a reason why I didn't use `attr_readonly` in the `Announcement` model (and resorted to checking things in a `before_save` – I think it's because it doesn't raise an error, so updates won't fail, but the change won't get persisted. I'd rather it errors hard (so we can catch bugs earlier). What do you think?

Reply:
Regarding `name` - I think we should stick with it being immutable. Is there any use case you have in mind for it? 
Re `attr_readonly` - Yes, that's right. It doesn't raise an error on update - it ignores those fields. I quite like that behaviour. The fact that it doesn't throw won't affect functionality in any way. API already filters out all non-modifiable fields and if you do update directly (e.g. from the console) it won't persist those changes. Can make it throw a dramatic error on update if you prefer that though :)

Reply:
Re: `name` – I was thinking of cases where a typo is made, or needing to change a `_` to a `-` (or vice versa) e.g. to make consistent with other names (like I think we've had to do before). Can keep it immutable and if we need to we'll make it modifiable later.

Re: `attr_readonly` – a "dramatic error" would be amazing ;-) My concern is running some commands in a console, or background job, or etc. and not realising that it hasn't been updated. But okay, let's keep as-is for now.